File: core/model/src/main/java/app/muplay/model/SubsonicCredentials.java
package app.muplay.model;

import javax.annotation.Nonnull;

/** Immutable Subsonic server credentials. */
public record SubsonicCredentials(
    @Nonnull String baseUrl, @Nonnull String username, @Nonnull String password) {

  public SubsonicCredentials {
    if (baseUrl.isBlank()) {
      throw new IllegalArgumentException("baseUrl must not be blank");
    }
    if (username.isBlank()) {
      throw new IllegalArgumentException("username must not be blank");
    }
    // Trailing slashes break path concatenation against /rest/*.
    baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
  }

  @Nonnull
  public static SubsonicCredentials create(
      @Nonnull String baseUrl, @Nonnull String username, @Nonnull String password) {
    return new SubsonicCredentials(baseUrl, username, password);
  }

  /** Never log credentials. */
  @Override
  @Nonnull
  public String toString() {
    return "SubsonicCredentials{" + username + "@" + baseUrl + "}";
  }
}

File: core/model/src/main/kotlin/app/muplay/model/SubsonicCredentials.kt
package app.muplay.model

/**
 * Credentials for a Subsonic-compatible server (e.g. Navidrome).
 *
 * [toString] is hand-written to omit [password]: `data class` generates a `toString()` that
 * includes every constructor property, and logging or crash-reporting a default one would leak
 * the plaintext password.
 */
data class SubsonicCredentials(
  val baseUrl: String,
  val username: String,
  val password: String,
) {
  override fun toString(): String =
    "SubsonicCredentials(baseUrl='$baseUrl', username='$username', password=<redacted>)"
}
